rmf_schedule_visualizer/src/rmf_schedule_visualizer/ScheduleDataNode.cpp

```diff
@@ -0,0 +1,258 @@
+namespace rmf_schedule_visualizer {
+
+//==============================================================================
+class ScheduleDataNode::Implementation
+{
+public:
+
+  using ConflictNotice = rmf_traffic_msgs::msg::NegotiationNotice;
+  using ConflictConclusion = rmf_traffic_msgs::msg::NegotiationConclusion;
+
+  Implementation() {}
+  Implementation(const Implementation&) {}
+
+  struct Data
+  {
+    rmf_traffic_ros2::schedule::MirrorManager mirror;
+
+    Data(rmf_traffic_ros2::schedule::MirrorManager mirror_)
+    : mirror(std::move(mirror_))
+    {
+      // Do nothing
+    }
+  };
+
+  rclcpp::Subscription<ConflictNotice>::SharedPtr conflict_notice_sub;
+  rclcpp::Subscription<ConflictConclusion>::SharedPtr conflict_conclusion_sub;
+
+  std::unique_ptr<Data> data;
+  mutable std::mutex mutex;
+  std::unordered_map<
+    rmf_traffic::schedule::Version,
+    std::vector<rmf_traffic::schedule::ParticipantId>> conflicts;
+
+  std::shared_ptr<rmf_traffic_ros2::schedule::Negotiation> negotiation =
+    nullptr;
+
+  void start(Data data);
+};
+
+//==============================================================================
+std::shared_ptr<ScheduleDataNode> ScheduleDataNode::make(
+  const std::string& node_name,
+  rmf_traffic::Duration wait_time)
+{
+  using ConflictNotice = rmf_traffic_msgs::msg::NegotiationNotice;
+  using ConflictConclusion = rmf_traffic_msgs::msg::NegotiationConclusion;
+
+  const auto start_time = std::chrono::steady_clock::now();
+  std::shared_ptr<ScheduleDataNode> schedule_data(
+    new ScheduleDataNode(std::move(node_name)));
+
+  schedule_data->_pimpl->conflict_notice_sub =
+    schedule_data->create_subscription<ConflictNotice>(
+      rmf_traffic_ros2::NegotiationNoticeTopicName,
+      rclcpp::QoS(10),
+      [schedule_data](ConflictNotice::UniquePtr msg)
+      {
+        std::lock_guard<std::mutex> guard(schedule_data->_pimpl->mutex);
+        schedule_data->_pimpl->conflicts[msg->conflict_version] =
+          msg->participants;
+      });
+
+  schedule_data->_pimpl->conflict_conclusion_sub =
+    schedule_data->create_subscription<ConflictConclusion>(
+      rmf_traffic_ros2::NegotiationConclusionTopicName,
+      rclcpp::ServicesQoS(),
+      [schedule_data](ConflictConclusion::UniquePtr msg)
+      {
+        std::lock_guard<std::mutex> guard(schedule_data->_pimpl->mutex);
+        schedule_data->_pimpl->conflicts.erase(msg->conflict_version);
+      });  
+
+  // Creating a mirror manager that queries over all
+  // Spacetime in the database schedule
+  auto mirror_mgr_future = rmf_traffic_ros2::schedule::make_mirror(
+    *schedule_data, rmf_traffic::schedule::query_all(),
+    &schedule_data->_pimpl->mutex);
+
+  const auto stop_time = start_time + wait_time;
+  while (rclcpp::ok() && std::chrono::steady_clock::now() < stop_time)
+  {
+    rclcpp::spin_some(schedule_data);
+    using namespace std::chrono_literals;
+    bool ready = (mirror_mgr_future.wait_for(0s) == std::future_status::ready);
```

Comment:
Why is the result of this comparison stored in a variable that's only used in the very next line?